src/raft/recv_install_snapshot.c

```diff
@@ -8,6 +8,350 @@
+ *
+ * - If a party notices that last known chunk index sent back to it
+ *   doesn't match it's own, the communication get's restarted from
+ *   the lowest known index.
+ *
+ * If a reply is not received the Leader will eventually timeout and retry
+ * sending the same message.
+ *
+ * ==Crashes of the Leader and Follower.
+ *
+ * Crashes of the Leader are handled by Raft when a new leader is elected
+ * and the snapshot process is restarted.
+ *
+ * If the Follower crashes, it will restart its state machine into the NORMAL
+ * state. If the Leader then sends a message which assumes the Follower is at
+ * the state prior to the crash, the Follower will reply using the message's
+ * result RPC using the failed=true flag. Upong receiving the message the
+ * Leader will restart the snapshot installation procedure.
+ *
+ * =State model
+ *
+ * Definitions:
+ *
+ * Rf -- raft index sent in AppendEntriesResult() from Follower to Leader
+ * Tf -- Follower's term sent in AppendEntriesResult() from Follower to Leader
+ *
+ * Tl -- Leader's term
+ * Rl -- raft index of the Leader
+ *
+  * Leader's state machine:
+ *
+ * +---------------------------------+
+ * |                                 |     AppendEntriesResult() received
+ * |        *Result(failed=true)     |     raft_log.find(Rf) == "FOUND"
+ * |        received                 |      +------------+
+ * |        +---------------> FOLLOWER_ONLINE <----------+
+ * |        |                        |
+ * |        |                        | AppendEntriesResult() received
+ * |        |                        V Rf << Rl && raft_log.find(Rf) == "ENOENTRY"
+ * |        +------------- FOLLOWER_NEEDS_SNAPSHOT
+ * |        |                        |
+ * |        |                        | InstallSnapshotResult() received
+ * |        |                        V
+ * |        +------------ SIGNATURES_CALC_STARTED  -------+ SignatureResult() received
+ * |        |                        |             <------+
+ * |        |                        | SignatureResult(done=true) received
+ * |        |                        V
+ * |        +----------- SNAPSHOT_INSTALLATION_STARTED
+ * |        |                        |
+ * |        |                        | CP_Result()/MV_Result() received
+ * |        |                        V
+ * |        +--------------- SNAPSHOT_CHUNCK_SENT -------+ CP_Result()/MV_Result()
+ * |        |                        |            <------+  received
+ * |        |                        | Raw snapshot iteration done
+ * |        |                        V
+ * |        +--------------- SNAPSHOT_DONE_SENT
+ * |                                 | InstallSnapshotResult() received
+ * +---------------------------------+
+ *
+ * Follower's state machine:
+ *
+ *                         +---------+
+ *                         |         | Signature/CP_Result/MV_Result received
+ *                         |         | *Result(failed=true) sent
```

Comment:
I'd expect that the trigger you mentioned shall be somewhere around `@` (`@ || %`). Also, could you think about rephrasing it? Maybe the following passage will give you an idea:
 (%) - Signature/CP_Result/MV_Result received and in the current state receiving of such message type is unexpected. Result(failed=true) sent.

BTW, `(@)` message has semantics of unexpected event (Leader re-elected) also.

Reply:
I was assuming the `failed` (now `unexpected`) flag would only be used to detect crashes on the follower side, hence why it was only present in the `NORMAL` state but you are right the mechanism has to be more flexible.

I have changed the diagram in the latest commit.